[FIX] 예약 기능 관련 버그 수정

kimdevv opened the pull request:
## Issue Number
closed #1116 

## As-Is
* [같은 멘티가 동일한 멘토링에 여러 번 예약을 할 수 있었습니다.](https://www.notion.so/bini-team/2becb79c55d080d8b14ef0bd48fc7653?source=copy_link#2becb79c55d080e5aa24ddba78f8e5d8)
* [승인된 상태의 예약을 거절 상태로 바꿀 수 있는 버그가 있었습니다.](https://www.notion.so/bini-team/2bdcb79c55d0802cb6c5c99173cb3ad1?source=copy_link)
* [멘티가 예약 신청 요청을 여러 번 보내면, 그만큼 예약이 중복 신청되었습니다.](https://www.notion.so/bini-team/2bdcb79c55d080fa8db2f558f051f2bc?source=copy_link)


## To-Be
* 특정 멘토링에 `APPROVED`, `PENDING` 상태의 요청이 존재하는 멘티는 동일한 멘토링에 중복 예약할 수 없습니다.
* 승인 상태의 요청은 거절 상태로 바꿀 수 없도록 하였습니다.


## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description
[예약 신청 시 연속으로 누르면 여러 번 예약이 신청되는 버그](https://www.notion.so/bini-team/2bdcb79c55d080fa8db2f558f051f2bc?source=copy_link)는 이번 PR에서 `특정 멘토링에 APPROVED, PENDING 상태의 요청이 존재하는 멘티는 동일한 멘토링에 중복 예약할 수 없습니다.` 기능을 추가함으로써 어느 정도 방지될 것이라 예상합니다.

하지만 해당 버그를 완벽하게 수정하지는 못할 것 같은데요, 그 이유는 아래와 같습니다.
1. 멘티 A가 멘토링 a에게 예약 1을 신청한다.
2. 예약 1 생성 요청이 서버로 날아가고 있는 상황이다. 서버에서는 아직 예약 1이 생성되지 않았다.
3. 예약 1이 아직 생성되지 않은 상태에서 멘티 A가 멘토링 a에게 예약 2를 중복 신청한다.
4. 서버에서는 아직 예약 1이 생성되지 않은 상황이라 멘티 A와 멘토링 a 사이에는 아무런 예약이 존재하지 않은 것으로 판단하여 예약 2도 생성한다.

이렇게 찰나의 순간에 빠르게 예약 신청 요청을 보내면 백엔드에서는 막을 수 있는 방법이 없을 것 같습니다. 프론트엔드의 도움도 필요할 것 같습니다. 예약 신청 버튼을 한 번 누르면 Alert 창이 뜨도록 한다던가...? 방법이 있다면 의견 남겨주세요.
@dlsxjzld @kimyou1102

kimdevv committed:
test: 예약 상태 변경 실패 테스트 추가

- 이미 처리된 예약에 대한 상태 변경 실패 테스트 추가
- 동일한 상태로 예약 상태 변경 시도에 대한 실패 테스트 추가
- 예약 상태 검증 로직에 승인 후 거절 상태 전환 방지 조건 추가
- 예약 상태 비교 시 동등성(equals) 대신 동일성(==) 체크로 변경

kimdevv committed:
feat: 중복 예약 방지 기능 추가

- 진행 중인 예약이 있는 경우 예약 생성 불가하도록 검증 로직 추가
- `existsByMentoringIdAndMenteeIdAndStatusIn` 메서드 추가로 예약 상태 존재 여부 확인
- 중복 예약 요청 시 예외를 발생시키는 `DuplicateReservationException` 추가
- 관련 예외 메시지를 `BusinessErrorMessage`에 추가
- 중복 예약 방지 테스트 케이스 2종 추가 (예약 및 진행 중인 멘토링 상황)

dlsxjzld commented:
프리가 "해줘" @kimdevv 


2번 눌러도 1번만 신청되도록 중복 요청을 막는 건 프론트엔드에서 처리 가능할 거 같아!
요청에 대한 로딩 상태를 둬서 요청을 보내는 도중에는 버튼 클릭을 막으면 될 거 같아!

jye326 commented on backend/fittoring/src/main/java/fittoring/domain/model/Reservation.java:
```diff
@@ -76,10 +76,10 @@ public void changeStatus(Status updateStatus) {
             throw new InvalidStatusException(BusinessErrorMessage.RESERVATION_STATUS_ALREADY_UPDATE.getMessage());
         }
-        if (this.status.equals(updateStatus)) {
+        if (this.status == updateStatus) {
```
Enum은 싱글톤이라 두 방식이 본질적으로 차이가 없지만, null 비교 시 equals일 때는 NPE가 발생하고 == 비교일 때는 false를 반환하네요!

jye326 commented on backend/fittoring/src/main/java/fittoring/domain/model/Reservation.java:
```diff
@@ -76,10 +76,10 @@ public void changeStatus(Status updateStatus) {
 
     private void validateReservation(Status updateStatus) {
-        if (this.status.isReject() || this.status.isComplete()) {
+        if (this.status.isReject() || this.status.isComplete() || (updateStatus.isReject() && this.status.isApprove())) {
```
p3: `(updateStatus.isReject() && this.status.isApprove())` 이미 승인된 예약을 거절하려고 시도하는 경우군요! 한번에 의미를 파악할 수 있도록 적절한 이름의 변수로 분리하는 것이 어떨까요?

dye0p replied:
아래쪽에 이미 만들어둔 메서드가 있는데 그걸 사용해도 좋겠네요~

jye326 commented on backend/fittoring/src/main/java/fittoring/domain/model/Reservation.java:
```diff
@@ -76,10 +76,10 @@ public void changeStatus(Status updateStatus) {
     }
 
     private void validateReservation(Status updateStatus) {
```
p2: 예약의 상태를 검증하는 메서드인데 `validateReservation`은 너무 포괄적인 의미 같아요! `validateReservationStatus` 로 구체화하면 어떨까요?

jye326 reviewed (commented):
수고하셨습니다! 리뷰 몇 개 남겼으니 확인 부탁드려요~

dye0p reviewed (approved):
고생하셨어요! 승인합니다~

> [예약 신청 시 연속으로 누르면 여러 번 예약이 신청되는 버그](https://www.notion.so/bini-team/2bdcb79c55d080fa8db2f558f051f2bc?source=copy_link)

이 문제 경쟁 상태로 보여요. 락을 거는 방법도 있을 것 같은데, 어떻게 생각하시나요!

kimdevv committed:
Merge branch 'develop' into feature/1116-fix-reservation

kimdevv committed:
refactor: 예약 상태 검증 로직 리팩토링

- 예약 상태 검증 메서드 이름을 명확하게 변경(validateReservation → validateReservationStatus)
- 상태 변경 검증 로직을 가독성 향상을 위해 메서드로 분리(canChangeStatus, isRejectingAfterApproved)
- 중복된 로직 제거 및 유지보수성을 향상시키도록 코드 재구성

kimdevv commented:
> 이 문제 경쟁 상태로 보여요. 락을 거는 방법도 있을 것 같은데, 어떻게 생각하시나요!

일종의 경쟁 상태가 맞는 것 같습니다. 비관적 락을 통해 삽입 시 쓰기 락을 걸고 예외 처리를 해주면 해결될 것 같아요!
그런데 프론트엔드에서 Alert 처리만 해주면 락을 걸지 않고도 해결할 수 있는 문제이기 때문에 성능상 락을 걸지 않았습니다. 프론트엔드 분들과 이야기해 봐야 할 것 같아요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/domain/model/Reservation.java:
```diff
@@ -71,19 +71,27 @@ public void changeStatusWithoutValidation(Status updateStatus) {
 
-    private void validateReservation(Status updateStatus) {
-        if (this.status.isReject() || this.status.isComplete()) {
+    private void validateReservationStatus(Status updateStatus) {
```
p2: 예약 대기 상태를 확인하는 `active_status` 컬럼을 추가하고 검증 처리를 도메인+유니크 제약조건으로 결합하면 Lock 없이 경쟁상태를 예방할 수 있을 것 같아요 (mentor, mentee, active_status)

UX로 막는 방법도 좋지만, 말씀해주신 부분은 따닥 이벤트만 방지할 수 있을 것 같아서 다른 예외상황을 고려해서 서버에서 강한 제약조건으로 막아주면 좋을 것 같아요!

> 프론트엔드에서 Alert 처리만 해주면 락을 걸지 않고도 해결할 수 있는 문제이기 때문에 성능상 락을 걸지 않았습니다.

ppparkta reviewed (commented):
고생하셨습니다. 경쟁상태 예방에 대한 부분 완전히 해소되면 바로 승인하겠습니다!!
많은 고민과 꼼꼼한 문서화 해주셔서 감사합니다 👍

kimdevv committed:
Merge branch 'develop' into feature/1116-fix-reservation
